### PR TITLE
Task/apd 36 vler endpoint

### PR DESCRIPTION
**Pull request by jnguy-va**
https://vasdvp.atlassian.net/browse/APD-31

Implementing Endpoint with VLER Direct: https://www.fhir.org/guides/argonaut/pd/StructureDefinition-argo-endpoint.html

### CODE REVIEW

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/location/LocationWrapper.java [hunk 1]
```diff
@@ -9,10 +9,9 @@
 @Value
 @Builder
 public final class LocationWrapper {
-  @Builder.Default
-  ProviderServicesResponse providerServicesResponse = ProviderServicesResponse.builder().build();
+  ProviderServicesResponse providerServicesResponse;
```
Comment (jjennings-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/location/LocationWrapper.java]: I would undo these Changes. They are causing null pointer exceptions/test breakage....

Next sprint, this will be completely refactored anyway as we plan to use NPI in place of PPMS

File: provider-directory/pom.xml [hunk 2]
```diff
@@ -8,7 +8,7 @@
     <relativePath/>
   </parent>
   <properties>
-    <jacoco.coverage>0.90</jacoco.coverage>
+    <jacoco.coverage>0.87</jacoco.coverage>
```
Comment (bschofield-va) [inline provider-directory/pom.xml]: :(

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+/** REST implementation of PPMS client. */
+@Slf4j
+@Component
+public class RestVlerClient implements VlerClient {
+
+  private final String baseUrl;
+
+  private final RestTemplate restTemplate;
+
+  @Value("${vler.pubkey}")
```
Comment (bschofield-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: Just observing a funny inconsistency... this is very short `vler.pubkey` and `vler.prvkey`, but this is all out, `vler.truststore.password`.

What are you really saving by abbreviating "public" and "private" into oblivion? Recommend using _whole_ words.

```
@Value("${vler.public-key}") String publicKey;
@Value("${vler.private-key}") String privateKey;
```

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                  .toUriString();
+          String dateString =
+              DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm:ss z").format(ZonedDateTime.now());
+          HttpHeaders headers = new HttpHeaders();
+          headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+          headers.setContentType(MediaType.APPLICATION_JSON);
+          headers.add("Authorization", authHeader(baseUrl, url, dateString));
+          headers.add("Date", dateString);
+          HttpEntity<?> requestEntity = new HttpEntity<>(headers);
+          System.out.println("RequestEntity: " + requestEntity);
```
Comment (bschofield-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: :(

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    }
+    if (response.error() != null && isNotBlank(response.error().message())) {
+      throw new VlerException(message + ", " + response.error().message());
+    }
+    return response;
+  }
+
+  @SneakyThrows
+  private String authHeader(String baseUrl, String url, String dateString) {
+    String endpoint = "/" + url.substring(url.indexOf(baseUrl) + baseUrl.length());
```
Comment (bschofield-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: No protection around indexing and substring. This method _assumes_ `url` contains `baseUrl`, but doesn't verify the assumption.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java [hunk 6]
```diff
@@ -1,3 +1,152 @@
...
+      @RequestParam("organization") String organization,
+      @RequestParam(value = "page", defaultValue = "1") @Min(1) int page,
+      @RequestParam(value = "_count", defaultValue = "15") @Min(0) int count) {
+    throw new UnsupportedOperationException();
+  }
+
+  /** Placeholder for search by Identifier. */
+  private Pair<List<EndpointWrapper>, Integer> searchIdentifier(
+      MultiValueMap<String, String> parameters) {
+    return null;
```
Comment (bschofield-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java]: ```throw new UnsupportedOperationException();```

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java [hunk 7]
```diff
@@ -1,3 +1,152 @@
...
+                          .build()))
+              .build());
+    }
+    return Pair.of(endpointWrapperPages, addressResponse.contacts().size());
+  }
+
+  /** Placeholder for search by Organization. */
+  private Pair<List<EndpointWrapper>, Integer> searchOrganization(
+      MultiValueMap<String, String> parameters) {
+    return null;
```
Comment (bschofield-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java]: `throw new UnsupportedOperationException();`?

File: provider-directory/src/main/resources/application.properties [hunk 8]
```diff
@@ -1,4 +1,9 @@
 ppms.url=unset
+
+vler.prvkey=unset
```
Comment (bschofield-va) [inline provider-directory/src/main/resources/application.properties]: :(

File: src/scripts/make-configs.sh [hunk 9]
```diff
@@ -14,8 +14,10 @@ Options
 
 Secrets Configuration
  This bash file is sourced and expected to set the following variables
- - KEYSTORE_PASSWORD
  - PPMS_URL
+ - VLER_PRV_KEY
```
Comment (bschofield-va) [inline src/scripts/make-configs.sh]: :(

File: src/scripts/make-configs.sh [hunk 10]
```diff
@@ -47,8 +49,10 @@ echo "Loading secrets: $SECRETS"
 . $SECRETS
 
 MISSING_SECRETS=false
-[ -z "$KEYSTORE_PASSWORD" ] && echo "Missing configuration: KEYSTORE_PASSWORD" && MISSING_SECRETS=true
 [ -z "$PPMS_URL" ] && echo "Missing configuration: PPMS_URL" && MISSING_SECRETS=true
+[ -z "$VLER_PRV_KEY" ] && echo "Missing configuration: VLER_PRV_KEY" && MISSING_SECRETS=true
```
Comment (bschofield-va) [inline src/scripts/make-configs.sh]: :(

File: src/scripts/make-configs.sh [hunk 11]
```diff
@@ -100,6 +104,9 @@ configValue provider-directory $PROFILE capability.contact.email "$(sendMoarSpam
 configValue provider-directory $PROFILE capability.security.token-endpoint https://fake.com/token
 configValue provider-directory $PROFILE capability.security.authorize-endpoint https://fake.com/authorize
 configValue provider-directory $PROFILE ppms.url "$PPMS_URL"
+configValue provider-directory $PROFILE vler.prvkey "$VLER_PRV_KEY"
```
Comment (bschofield-va) [inline src/scripts/make-configs.sh]: :(

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        });
+  }
+
+  /** Acceptance of SSL for Java 12. */
+  @SneakyThrows
+  public RestTemplate restTemplate() {
+    TrustStrategy acceptingTrustStrategy = (X509Certificate[] chain, String authType) -> true;
+    SSLContext sslContext =
+        org.apache.http.ssl.SSLContexts.custom()
+            .loadTrustMaterial(null, acceptingTrustStrategy)
```
Comment (bschofield-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: Documentation is sketchy in `SSLContexts`, but looking at the code `.loadTrustMaterial(null, acceptingTrustStrategy)` and `.loadTrustMaterial(acceptingTrustStrategy)` do the same.

Docs do say this about `TrustStrategy`
> A strategy to establish trustworthiness of certificates without consulting the trust manager configured in the actual SSL context. This interface can be used to override the standard JSSE certificate verification process.

The `(...) -> true` function, combined with not actually loading a trust store effectively disables all SSL checks on this end of the mutual TLS side. Is that what you intended?

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+  public RestVlerClient(
+      @Value("${vler.url}") String baseUrl, @Autowired RestTemplate restTemplate) {
+    this.baseUrl = baseUrl.endsWith("/") ? baseUrl : baseUrl + "/";
+    this.restTemplate = restTemplate;
+  }
+
+  /*  Implement this if the above isnt secure enough.
+  @SneakyThrows
+  private RestTemplate secureRestTemplate() {
```
Comment (bschofield-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: These seems better from a security perspective.

**Comment by jblefkove-va** [comment]
Haven't done a full review yet, but two general comments:

1. I agree with Bryan that disabling SSL altogether is not the right approach. I think we should follow what hand-of-queen-elizabeth uses for E&E calls:
https://github.com/department-of-veterans-affairs/health-apis-eligibility-and-enrollment-client/blob/master/hand-of-queen-elizabeth/src/main/java/gov/va/api/health/queenelizabeth/ee/impl/ConnectionProvider.java#L65

Note that you don't have to specify the full path to the truststore. For example, in CCE we set the property `ee.truststore.path=eligibilityandenrollment-nonprod-truststore.jks`. The keystore just needs to be available in `src/main/resources`.

2. I don't know how long it's going to take to debug VLER's address filtering, especially since the issue may be on their side. In the interim, we should do our own filtering on the records we get back from VLER so that the results are still correct... and document why we're doing that.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/AddressResponse.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+  @Data
+  @Builder
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  public static final class Contacts {
+    /* Map to 'name'. */
+    @JsonProperty("displayname") // Last, First
+    private String displayname;
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/AddressResponse.java]: This should be `displayName`. We have to match VLER's field names in the `@JsonProperty()` annotations, but we can still name our fields whatever we want.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/AddressResponse.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+  @Data
+  @Builder
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  public static final class Contacts {
+    /* Map to 'name'. */
+    @JsonProperty("displayname") // Last, First
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/AddressResponse.java]: Our formatter sometimes pukes on trailing comments. Move this and the other trailing comments up into the corresponding javadoc instead.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/AddressResponse.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    private String mail;
+
+    @JsonProperty("uid") // LH of email address
+    private String uid;
+
+    @JsonProperty("givenname") // First name
+    private String givenname;
+
+    @JsonProperty("sn") // Surname
+    private String sn;
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/AddressResponse.java]: Since we can name our fields as we see fit, in this case I would rename the field to `surname` and delete the comment.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/VlerResponse.java [hunk 17]
```diff
@@ -0,0 +1,23 @@
+package gov.va.api.health.providerdirectory.service;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+public interface VlerResponse {
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/VlerResponse.java]: If the `error` body on VLER messages isn't applicable, we can remove this interface.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+  @Value("${vler.truststore.location}")
+  String truststoreLocation;
+
+  @Value("${vler.url}")
+  String vlerUrl;
+
+  public RestVlerClient(
+      @Value("${vler.url}") String baseUrl, @Autowired RestTemplate restTemplate) {
+    this.baseUrl = baseUrl.endsWith("/") ? baseUrl : baseUrl + "/";
+    this.restTemplate = restTemplate;
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: This is awkward in that we're not using the `restTemplate` that is autowired in. So let's remove the autowired one and initialize `restTemplate` ourselves here, something like `this.restTemplate = secureRestTemplate();`

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+  String truststorePassword;
+
+  @Value("${vler.truststore.location}")
+  String truststoreLocation;
+
+  @Value("${vler.url}")
+  String vlerUrl;
+
+  public RestVlerClient(
+      @Value("${vler.url}") String baseUrl, @Autowired RestTemplate restTemplate) {
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: For the sake of consistency, let's add all of the fields to this constructor and move the `@Value()` annotations down here.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/VlerClient.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+/** This is the abstraction for communicating with PPMS through their Restful API. */
+public interface VlerClient {
+
+  AddressResponse endpointByAddress(String address);
+
+  /** A request was malformed, such as missing required search parameters. */
+  public static final class BadRequest extends VlerException {
+
+    public BadRequest(String message, Throwable cause) {
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/VlerClient.java]: Are these all the same exceptions that were defined for PPMS? Seems like we should consolidate these into a shared `Exceptions` class

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java [hunk 21]
```diff
@@ -1,3 +1,152 @@
...
+    int toIndex = Math.min((fromIndex + count), addressResponse.contacts().size());
+    List<AddressResponse.Contacts> addressResponsePages =
+        addressResponse.contacts().subList(fromIndex, toIndex);
+    List<EndpointWrapper> endpointWrapperPages = new ArrayList<>();
+    for (int i = 0; i < addressResponsePages.size(); i++) {
+      endpointWrapperPages.add(
+          (EndpointWrapper.builder()
+                  .addressResponse(
+                      AddressResponse.builder()
+                          .contacts(singletonList(addressResponsePages.get(i)))
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java]: I made a similar comment on a previous PR, but it is unnecessarily complicated here to create this fake `AddressResponse` that has a singleton list containing `addressResponsePages.get(i)`. It is more straightforward (both here and in the transformer later) to change `EndpointWrapper` to have a field that is just the single `addressResponsePages.get(i)` value instead.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointTransformer.java [hunk 22]
```diff
@@ -1,3 +1,51 @@
...
+    }
+    List<Coding> codings = new ArrayList<>();
+    codings.add(Coding.builder().display(payloadTypeResponse).code(payloadTypeResponse).build());
+    List<CodeableConcept> payload = new ArrayList<>();
+    payload.add(CodeableConcept.builder().coding(codings).build());
+    return payload;
+  }
+
+  Endpoint.Status status(String status) {
+    return Endpoint.Status.active;
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointTransformer.java]: This method isn't doing anything, so let's remove it and just hardcode `Endpoint.Status.active` where it is used.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointTransformer.java [hunk 23]
```diff
@@ -1,3 +1,51 @@
...
+        .name(response.displayname())
+        .payloadType(payloadCodeCodings("VLER Direct"))
+        .address(response.mail())
+        .build();
+  }
+
+  Coding connectionCodeCodings(String connectionTypeResponse) {
+    if (connectionTypeResponse == null) {
+      return null;
+    }
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointTransformer.java]: Similar to previous comment... if this method is only ever invoked with `direct-project` as the argument, then you don't need the null check. You don't really need the method either; just inline `Coding.builder()....` where it is used.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointTransformer.java [hunk 24]
```diff
@@ -1,3 +1,51 @@
...
+    Coding vlerDefault = Coding.builder().code(connectionTypeResponse).build();
+    return vlerDefault;
+  }
+
+  List<CodeableConcept> payloadCodeCodings(String payloadTypeResponse) {
+    if (payloadTypeResponse == null) {
+      return null;
+    }
+    List<Coding> codings = new ArrayList<>();
+    codings.add(Coding.builder().display(payloadTypeResponse).code(payloadTypeResponse).build());
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointTransformer.java]: You can create and populate the list in one line with `Arrays.asList(Coding.builder()...)`

**Comment by jblefkove-va** [comment]
Also, as part of the above change, update the `make-configs` script to set the truststore name and password properties as well.

**Comment by bschofield-va** [comment]
I would not recommend placing the keystore _inside_ the application. This will mean you'll need to modify the source tree, rebuild, and redeploy when the cert expires. I suggest externalizing the cert as part of the deployment unit and bringing it into the application via s3 copy that happens when the pod starts up.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/VlerResponse.java (hunk 17 above)
Comment (jnguy-va) [reply provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/VlerResponse.java]: Error body removed.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+    return response;
+  }
+
+  @SneakyThrows
+  private String authHeader(String baseUrl, String url, String dateString) {
+    String endpoint;
+    try {
+      endpoint = "/" + url.substring(url.indexOf(baseUrl) + baseUrl.length());
+    } catch (Exception e) {
+      throw new BadRequest("Base URL not found within url ", e);
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: The error message should include the values of `baseUrl` and `url`

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+
+  /** Calls the VLER Direct API by email address search. */
+  @Override
+  public AddressResponse endpointByAddress(String address) {
+    return handleVlerExceptions(
+        address,
+        () -> {
+          String url =
+              UriComponentsBuilder.fromHttpUrl(baseUrl + "/direct/addresses")
+                  .queryParam("cn", address)
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: Can we remove this parameter since it's not really doing anything?

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+                  .toUriString();
+          String dateString =
+              DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm:ss z").format(ZonedDateTime.now());
+          HttpHeaders headers = new HttpHeaders();
+          headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+          headers.setContentType(MediaType.APPLICATION_JSON);
+          headers.add("Authorization", authHeader(baseUrl, url, dateString));
+          headers.add("Date", dateString);
+          ResponseEntity<AddressResponse> entity =
+              secureRestTemplate()
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: This should be `restTemplate`, the field.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+                  .exchange(url, HttpMethod.GET, new HttpEntity<>(headers), AddressResponse.class);
+          return entity.getBody();
+        });
+  }
+
+  @SneakyThrows
+  private RestTemplate secureRestTemplate() {
+    ClassLoader cl = getClass().getClassLoader();
+    if (cl == null) {
+      throw new ClassLoaderException("Something went wrong getting the class loader");
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: I know this from queen-elizabeth, but I think the dedicated `ClassLoaderException` class is overkill. Let's remove that and just throw a more generic `RuntimeException` instead.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/VlerClient.java [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+      super(message + ", " + cause.getMessage(), cause);
+    }
+  }
+
+  /** The resource requested was not found. */
+  public static final class NotFound extends VlerException {
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/VlerClient.java]: Above comment still stands... the `NotFound`, `SearchFailed` and `BadRequest` exceptions can be refactored into another class and shared with `RestPpmsClient`. (The PPMS exceptions inherit from `PpmsException`, while these inherit from `VlerException`. I'm inclined to say that both parents `PpmsException` and `VlerException` can be deleted, and the `NotFound`, `SearchFailed` and `BadRequest` exceptions can extend from `RuntimeException` directly.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java [hunk 30]
```diff
@@ -1,3 +1,165 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Request Mappings for Endpoint Resource, see
+ * http://www.fhir.org/guides/argonaut/pd/StructureDefinition-argo-endpoint.html for implementation
+ * details.
+ */
+@RestController
+@RequestMapping(
+  value = {"/api/Endpoint"},
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java]: This should be just `/Endpoint` now. Josh changed the other instances in his PR this afternoon.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java [hunk 31]
```diff
@@ -1,3 +1,165 @@
...
+            .totalRecords(root.getRight())
+            .build();
+    return bundler.bundle(
+        BundleContext.of(
+            linkConfig, root.getLeft(), transformer, Endpoint.Entry::new, Endpoint.Bundle::new));
+  }
+
+  /** Read by identifier. */
+  @GetMapping(value = {"/{publicId}"})
+  public Endpoint readByIdentifier(@PathVariable("publicId") String publicId) {
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java]: In data-query you see `publicId` because there is translation between public-facing IDs and CDW IDs. Since we're not doing any ID translation, this can just be 'id'.

File: provider-directory/src/main/resources/application.properties [hunk 32]
```diff
@@ -1,4 +1,7 @@
 ppms.url=unset
+vler.key.public=unset
+vler.key.private=unset
+vler.url=unset
```
Comment (jblefkove-va) [inline provider-directory/src/main/resources/application.properties]: For the make-configs script to work, I think the two trust-store variables need to be `unset` in here as well.

File: src/scripts/make-configs.sh [hunk 33]
```diff
@@ -16,6 +16,9 @@ Secrets Configuration
  This bash file is sourced and expected to set the following variables
  - KEYSTORE_PASSWORD
  - PPMS_URL
+ - VLER_KEY_PUBLIC
+ - VLER_KEY_PRIVATE
+ - VLER_URL
```
Comment (jblefkove-va) [inline src/scripts/make-configs.sh]: Add the two truststore variables to this script.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 34]
```diff
@@ -0,0 +1,172 @@
...
+        cl.getResourceAsStream(FilenameUtils.getName(vlerTruststore))) {
+      KeyStore ts = KeyStore.getInstance("JKS");
+      ts.load(truststoreInputStream, vlerTruststorePassword.toCharArray());
+      SSLContext sslContext = null;
+      TrustManagerFactory trustManagerFactory =
+          TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+      trustManagerFactory.init(ts);
+      sslContext = SSLContext.getInstance("TLSv1.1");
+      sslContext.init(null, trustManagerFactory.getTrustManagers(), new SecureRandom());
+      HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory());
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: I don't have my GFE at the moment to confirm, but I think these three lines here (configuring SSL context and then the `HttpsURLConnection` socket factory) are unnecessary. I think you just need the trust store and the Spring `SSLConnectionSocketFactory` below.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java [hunk 35]
```diff
@@ -0,0 +1,172 @@
...
+              DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm:ss z").format(ZonedDateTime.now());
+          HttpHeaders headers = new HttpHeaders();
+          headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+          headers.setContentType(MediaType.APPLICATION_JSON);
+          headers.add("Authorization", authHeader(baseUrl, url, dateString));
+          headers.add("Date", dateString);
+          ResponseEntity<AddressResponse> entity =
+              secureRestTemplate()
+                  .exchange(url, HttpMethod.GET, new HttpEntity<>(headers), AddressResponse.class);
+          return entity.getBody();
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/client/RestVlerClient.java]: After we make the call, I think we should do the filtering (that is currently done in `searchName` in the controller class) here. That way, from the perspective of the rest of the code, `VlerClient.endpointByAddress()` still has the expected behavior of returning just the results for whatever search parameter was passed in.

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java [hunk 36]
```diff
@@ -1,3 +1,165 @@
...
+  }
+
+  /** Logic for search using full body from VLER. */
+  private Pair<List<EndpointWrapper>, Integer> searchName(
+      MultiValueMap<String, String> parameters, String function) {
+    String searchFunction = parameters.getFirst(function);
+    AddressResponse addressResponse = vlerClient.endpointByAddress(searchFunction);
+    int filteredCount = 0;
+    List<AddressResponse.Contacts> unfilteredAddressResponsePages =
+        addressResponse.contacts().subList(0, addressResponse.contacts().size());
```
Comment (jblefkove-va) [inline provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java]: It looks like we lost the paging here? Although in practice I doubt there would ever be enough results for us to have multiple pages.

File: provider-directory/src/main/resources/application.properties (hunk 32 above)
Comment (jnguy-va) [reply provider-directory/src/main/resources/application.properties]: Placed them under this comment spot

File: src/scripts/make-configs.sh (hunk 33 above)
Comment (jnguy-va) [reply src/scripts/make-configs.sh]: Also placed under this comment spot

File: provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java (hunk 36 above)
Comment (jnguy-va) [reply provider-directory/src/main/java/gov/va/api/health/providerdirectory/service/controller/endpoint/EndpointController.java]: Paging added!  Was very broken indeed
